Std/Data/List/Perm.lean

```diff
@@ -0,0 +1,852 @@
+  constructor
+
+theorem perm_insertNth {α} (x : α) (l : List α) {n} (h : n ≤ l.length) :
+    insertNth n x l ~ x :: l := by
+  induction l generalizing n with
+  | nil =>
+    cases n
+    . apply Perm.refl
+    . cases h
+  | cons _ _ l_ih =>
+    cases n
+    · simp [insertNth]
+    · simp only [insertNth, modifyNthTail]
+      refine' Perm.trans (Perm.cons _ (l_ih _)) _
+      · apply Nat.le_of_succ_le_succ h
+      · apply Perm.swap
+
+theorem Perm.union_right {l₁ l₂ : List α} (t₁ : List α) (h : l₁ ~ l₂) : l₁ ∪ t₁ ~ l₂ ∪ t₁ := by
+  induction h with
+  | nil         => apply Perm.refl
+  | cons a _ ih => exact ih.insert a
+  | swap        => apply perm_insert_swap
+  | trans _ _ ih_1 ih_2 => exact ih_1.trans ih_2
+
+theorem Perm.union_left (l : List α) {t₁ t₂ : List α} (h : t₁ ~ t₂) : l ∪ t₁ ~ l ∪ t₂ := by
+  induction l with
+  | nil => simp only [nil_union, h]
+  | cons _ _ ih => simp only [cons_union, Perm.insert _ ih]
+
+theorem Perm.union {l₁ l₂ t₁ t₂ : List α} (p₁ : l₁ ~ l₂) (p₂ : t₁ ~ t₂) :
+    l₁ ∪ t₁ ~ l₂ ∪ t₂ :=
+  (p₁.union_right t₁).trans (p₂.union_left l₂)
+
+theorem Perm.inter_right {l₁ l₂ : List α} (t₁ : List α) : l₁ ~ l₂ → l₁ ∩ t₁ ~ l₂ ∩ t₁ :=
+  Perm.filter _
+
+theorem Perm.inter_left (l : List α) {t₁ t₂ : List α} (p : t₁ ~ t₂) : l ∩ t₁ = l ∩ t₂ :=
+  filter_congr' fun a _ => by
+    have := p.mem_iff (a := a)
+    simpa using this
+
+theorem Perm.inter {l₁ l₂ t₁ t₂ : List α} (p₁ : l₁ ~ l₂) (p₂ : t₁ ~ t₂) : l₁ ∩ t₁ ~ l₂ ∩ t₂ :=
+  p₂.inter_left l₂ ▸ p₁.inter_right t₁
+
+end
+
+theorem Perm.pairwise_iff {R : α → α → Prop} (S : ∀ {x y}, R x y → R y x) :
+    ∀ {l₁ l₂ : List α} (_p : l₁ ~ l₂), Pairwise R l₁ ↔ Pairwise R l₂ :=
+  suffices ∀ {l₁ l₂}, l₁ ~ l₂ → Pairwise R l₁ → Pairwise R l₂
+    from fun p => ⟨this p, this p.symm⟩
+  @fun l₁ l₂ p d => by
+  induction d generalizing l₂ with
+  | nil =>
+    rw [← p.nil_eq]
+    constructor
+  | cons h _ IH =>
+    have : _ ∈ l₂ := p.subset (mem_cons_self _ _)
+    rcases append_of_mem this with ⟨s₂, t₂, rfl⟩
+    have p' := (p.trans perm_middle).cons_inv
+    refine' (pairwise_middle S).2 (pairwise_cons.2 ⟨fun b m => _, IH _ p'⟩)
+    exact h _ (p'.symm.subset m)
+
+theorem Pairwise.perm {R : α → α → Prop} {l l' : List α} (hR : l.Pairwise R) (hl : l ~ l')
+    (hsymm : ∀ {x y}, R x y → R y x) : l'.Pairwise R :=
+  (hl.pairwise_iff hsymm).mp hR
+
+theorem Perm.pairwise {R : α → α → Prop} {l l' : List α} (hl : l ~ l') (hR : l.Pairwise R)
+    (hsymm : ∀ {x y}, R x y → R y x) : l'.Pairwise R :=
+  hR.perm hl hsymm
+
+theorem Perm.nodup_iff {l₁ l₂ : List α} : l₁ ~ l₂ → (Nodup l₁ ↔ Nodup l₂) :=
+  Perm.pairwise_iff <| @Ne.symm α
+
+theorem Perm.join {l₁ l₂ : List (List α)} (h : l₁ ~ l₂) : l₁.join ~ l₂.join :=
+  Perm.recOn h (Perm.refl _) (fun x xs₁ xs₂ _ ih => ih.append_left x)
+    (fun x₁ x₂ xs => by
+      simp [join]
+      rw [←append_assoc, ←append_assoc]
+      refine perm_append_comm.append_right ?_)
+    @fun xs₁ xs₂ xs₃ _ _ => Perm.trans
```

Comment:
```suggestion
theorem Perm.join {l₁ l₂ : List (List α)} (h : l₁ ~ l₂) : l₁.join ~ l₂.join :=
  Perm.recOn h (Perm.refl _) (fun x xs₁ xs₂ _ ih => ih.append_left x)
    (fun x₁ x₂ xs => by
      simp [join]
      rw [←append_assoc, ←append_assoc]
      refine perm_append_comm.append_right ?_)
    @fun xs₁ xs₂ xs₃ _ _ => Perm.trans
```

Reply:
I think GH messed up this suggested change?

Reply:
(or at least the suggestion seems identical to the current diff?)

Reply:
Looks like copypasta. It probably was just another "use induction instead of rec" but it's lost now...

Reply:
Found it! Fixed now.